### PR TITLE
PUBDEV-8900: Estimate tweedie dispersion parameter, variance power using maximum likelihood.

### PR DESCRIPTION
**Pull request by tomasfryda**
https://h2oai.atlassian.net/browse/PUBDEV-8900

This PR contains some improvements to the original `TweedieEstimator` that make it more likely that the Fourier inversion method will terminate - there are some rare cases caused by rounding that cause increment in finding zeros to be so small so that is rounded to zero. This should be fixed and it shouldn't be concern for the variance power estimation (these cases occur with p very close to 1 (e.g. 1+1e-16) and at the same time phi very close to 0 (e.g., 1e-16)) or the estimation of both `phi` and `p` (as both optimization procedures limit the ranges of those parameters). However, it could cause issue when calculating AIC with user specified variance power and dispersion.

Another improvement is to make sure that `init()` is allowed to complete in order to show the model builder errors - now in master when used doesn't specify `tweedie_variance_power` and has dispersion method equal to `ml` the `init` method will try to build null model and it will fail since there is an `assert` in the `TweedieEstimator` that makes sure that `p>1`.

This PR contains implementation of Nelder-Mead optimization method[1,2] with one modification it is constrained so that $p \geq 1+10^{-10}$  and $\phi \geq 10^{-10}$.

[1] https://en.wikipedia.org/wiki/Nelder%E2%80%93Mead_method
[2] https://people.duke.edu/~hpgavin/cee201/Nelder-Mead-2D.pdf

### CODE REVIEW

**Review by wendycwong — CHANGES_REQUESTED** [review]

**Comment by wendycwong** [comment]
If you run this code:

setwd(normalizePath(dirname(R.utils::commandArgs(asValues=TRUE)$"f")))
source("../../../scripts/h2o-r-test-setup.R")
##
# In this test, we are going to generate real tweedie dataset, run H2O GLM and R GLM and compare the results for
# accuracy and others
##

test_glm_tweedies <- function() {
  if (requireNamespace("tweedie")) {
    num_rows <- 10000
    num_cols <- 5
    f1 <- random_dataset_real_only(num_rows, num_cols) # generate dataset containing the predictors.
    f1R <- as.data.frame(h2o.abs(f1))
    weights <- c(0.1, 0.2, 0.3, 0.4, 0.5, 1) # weights to generate the mean
    mu <- generate_mean(f1R, num_rows, num_cols, weights)
    pow <- c(1.1, 1.1, 1.1, 1.5, 1.5, 1.5, 1.9, 1.9, 1.9, 2.1, 2.1, 2.1, 2.5, 2.5, 2.5, 10, 10, 10) # variance power range
    phi <- c(0.1, 1, 10, 0.1, 1, 10, 0.1, 1, 10, 0.1, 1, 10, 0.1, 1, 10, 0.1, 1, 10)   # dispersion factor range
    y <- "resp"      # response column
    x <- c("abs.C1.", "abs.C2.", "abs.C3.", "abs.C4.", "abs.C5.")
    for (ind in c(1:length(pow))) { # generate dataset with each variance power and dispersion factor
      trainF <- generate_dataset(f1R, num_rows, num_cols, pow[ind], phi[ind], mu)
      checkOutH2O(pow[ind], 1-pow[ind], x, y, trainF, as.data.frame(trainF), phi[ind])
    }
  } else {
    print("test_glm_tweedies is skipped.  Need to install tweedie package.")
  }
}

generate_dataset<-function(f1R, numRows, numCols, pow, phi, mu) {
  # resp <- c(1:numRows)
  # for (rowIndex in c(1:numRows)) {
  #   resp[rowIndex] <- tweedie::rtweedie(1,xi=pow,mu[rowIndex], phi, power=pow)
  # }
  resp <- tweedie::rtweedie(numRows, xi=pow, mu, phi, power=pow)
  f1h2o <- as.h2o.data.frame(f1R)
  resph2o <- as.h2o.data.frame(as.data.frame(resp))
  finalFrame <- h2o.cbind(f1h2o, resph2o)
  return(finalFrame)
}

generate_mean<-function(f1R, numRows, numCols, weights) {
  y <- c(1:numRows)
  for (rowIndex in c(1:numRows)) {
    tempResp = 0.0
    for (colIndex in c(1:numCols)) {
      tempResp = tempResp+weights[colIndex]*f1R[rowIndex, colIndex]
    }
    y[rowIndex] = tempResp
  }
  return(y)
}

checkOutH2O <-
  function(vpower,
           lpower,
           x,
           y,
           hdf,
           df,
           truedisp,
           tolerance = 2e-4) {
    print("Fixing variance power and just estimate dispersion parameters")
    h2omodel1 <-
      h2o.glm(
        x = x,
        y = y,
        training_frame = hdf,
        family = "tweedie",
        link = "tweedie",
        tweedie_variance_power = vpower,
        tweedie_link_power = lpower,
        alpha = 0.5,
        lambda = 0,
        nfolds = 0,
        dispersion_parameter_method="ml",
        fix_dispersion_parameter=FALSE,
        fix_tweedie_variance_power=TRUE,
        compute_p_values = TRUE
      )
    printDispersion(h2omodel1, vpower, truedisp)
    print("Fixing dispersion parameters and just estimate variance power")
    h2omodel2 <-
      h2o.glm(
        x = x,
        y = y,
        training_frame = hdf,
        family = "tweedie",
        link = "tweedie",
        tweedie_variance_power = vpower,
        tweedie_link_power = lpower,
        alpha = 0.5,
        lambda = 0,
        nfolds = 0,
        init_dispersion_parameter = truedisp,
        dispersion_parameter_method="ml",
        fix_dispersion_parameter=TRUE,
        fix_tweedie_variance_power=FALSE,
        compute_p_values = TRUE
      )
    printVpower(h2oModel2, vpower, trueDisp)
    print("Estimate both dispersion and variance power")
    h2omodel3 <-
      h2o.glm(
        x = x,
        y = y,
        training_frame = hdf,
        family = "tweedie",
        link = "tweedie",
        tweedie_variance_power = vpower,
        tweedie_link_power = lpower,
        alpha = 0.5,
        lambda = 0,
        nfolds = 0,
        init_dispersion_parameter = truedisp,
        dispersion_parameter_method="ml",
        fix_dispersion_parameter=TRUE,
        fix_tweedie_variance_power=TRUE,
        compute_p_values = TRUE
      )
    printDispersion(h2oModel3, vpower, trueDisp)
    printVpower(h2oModel3, vpower, trueDisp)
  }

printDispersion <- function(h2oModel, vPower, trueDisp) {
  print(past("true variance power ", vPower))
  print(paste("true dispersion parameter ", truedisp))
  print(paste(
    "H2O model dispersion estimate",
    h2oModel@parameters$init_dispersion_parameter,
    sep = ":"
  ))
}

printVpower <- function(h2oModel, vPower, trueDisp) {
  print(past("true dispersion power ", trueDisp))
  print(paste("true variance power ", vpower))
  print(paste("H2O model variance power estimation", h2oModel@parameters$tweedie_variance_power))
}



doTest("Comparison of H2O to R TWEEDIE family coefficients and disperson with tweedie dataset", test_glm_tweedies)

It gets stuck right at the beginning for some reason.  I was just having it estimate dispersion with variance power fixed.....  Can you take a look?

**Comment by wendycwong** [comment]
Tomas:  This is so embarrassing.  I have many bugs in my code that I sent you .  You will need to fix them before you can run them.

**Review by wendycwong — APPROVED** [review]

**Review by wendycwong — APPROVED** [review]

**Review by wendycwong — APPROVED** [review]

**Review by wendycwong — APPROVED** [review]

**Review by wendycwong — APPROVED** [review]

**Review by wendycwong — APPROVED** [review]

**Review by wendycwong — APPROVED** [review]

**Review by wendycwong — APPROVED** [review]

**Comment by tomasfryda** [comment]
I modified the selection of the initial points to use the information from the previous estimation:
- make a triangle with the center that is the previous estimate
- I didn't know how to pick the triangle so I rotate it around the center so that it's the same every 5th iteration (by rotating it by $\frac{2\pi}{15}$) this made it possible to use smaller area for the simplex without getting stuck in local optimum => speedup by the factor of ~2
- I also enable the user to influence the size of this rotated simplex by changing the `dispersion_learning_rate` as it could be handy to increase it in case it gets stuck in local optimum

I tried to use the model builder warn but it didn't work for me. My understanding is that I would need to warn from init.
From debugging R's client I think it works roughly like this:
```mermaid
sequenceDiagram
R/Python client->>H2O backend: Hello, could you build an h2o glm with these params?
Note right of H2O backend: ModelBuilder.init
H2O backend->>R/Python client: Sure. Here's the key and some warnings.
Note right of H2O backend: Building a model...
loop while not model ready
    R/Python client->>H2O backend: Is the model ready?
end

R/Python client->>H2O backend: h2o.get_model(key)
```

Python seems to have the similar logic.
So I return `NaN` in case the dispersion diverges and warn using `Log.warn`.
